File: generator/plugins/express/templates/src/seed.js
'use strict';

// Seeds the default login user once, on startup, if it does not already exist.
// The password is hashed here at RUNTIME (with a per-install salt) — this is
// application behaviour, not generator output, so it does not affect the
// deterministic, byte-for-byte generation guarantee (ADR-003).
const bcrypt = require('bcryptjs');
const pool = require('./db');

async function seed() {
  const username = process.env.APP_SEED_ADMIN_USERNAME || 'admin';
  const password = process.env.APP_SEED_ADMIN_PASSWORD || 'admin123';
  const { rowCount } = await pool.query('SELECT 1 FROM users WHERE username = $1', [username]);
  if (rowCount === 0) {
    const hash = bcrypt.hashSync(password, 10);
    await pool.query('INSERT INTO users (username, password_hash, enabled) VALUES ($1, $2, TRUE)', [
      username,
      hash,
    ]);
    console.log('Seeded default user "' + username + '"');
  }
}

module.exports = seed;

File: generator/plugins/go/templates/internal/seed/seed.go
// THRAKSHA-owned shell. Seeds the default login user once, on startup.
//
// The password is hashed here at RUNTIME (with a per-install salt) — application
// behaviour, not generator output — so it does not affect the deterministic,
// byte-for-byte generation guarantee (ADR-003).
package seed

import (
	"database/sql"
	"log"
	"os"

	"golang.org/x/crypto/bcrypt"
)

// Run inserts the default admin user if it does not already exist.
func Run(db *sql.DB) error {
	username := getenv("APP_SEED_ADMIN_USERNAME", "admin")
	password := getenv("APP_SEED_ADMIN_PASSWORD", "admin123")

	var one int
	err := db.QueryRow("SELECT 1 FROM users WHERE username = __DB_PH1__", username).Scan(&one)
	if err == nil {
		return nil // already seeded
	}
	if err != sql.ErrNoRows {
		return err
	}

	// bcrypt only considers the first 72 bytes; truncate to stay within its limit.
	pw := []byte(password)
	if len(pw) > 72 {
		pw = pw[:72]
	}
	hash, err := bcrypt.GenerateFromPassword(pw, bcrypt.DefaultCost)
	if err != nil {
		return err
	}
	if _, err := db.Exec(
		"INSERT INTO users (username, password_hash, enabled) VALUES (__DB_PH1__, __DB_PH2__, TRUE)",
		username, string(hash),
	); err != nil {
		return err
	}
	log.Printf("Seeded default user %q", username)
	return nil
}

func getenv(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}
